alibi/explainers/ale.py

```diff
@@ -0,0 +1,479 @@
+    """
+
+    def minimum_satisfied(n: int) -> int:
+        """
+        Calculates whether the partition into n quantiles has the minimum number
+        of points in each resulting bin.
+        """
+        q = np.unique(get_quantiles(values, num_points=n))
+        indices = np.searchsorted(q, values, side='left')
+        indices[indices == 0] = 1
+        interval_n = np.bincount(indices)
+        return int(np.all(interval_n[1:] > min_bin_points))
+
+    # bisect
+    num_points = bisect_fun(fun=lambda x: 1 - minimum_satisfied(x), target=0.5, lo=0, hi=len(values)) - 1
+    assert minimum_satisfied(num_points)
+    q = np.unique(get_quantiles(values, num_points=num_points))
+
+    return q, num_points
+
+
+def ale_num(
+        predictor: Callable,
+        X: np.ndarray,
+        feature: int,
+        min_bin_points: int = 4) -> Tuple[np.ndarray, ...]:
+    """
+    Calculate the first order ALE curve for a numerical feature.
+
+    Parameters
+    ----------
+    predictor
+        Model prediction function.
+    X
+        Dataset used to calculate the ALE curves.
+    feature
+        Index of the numerical feature for which to calculate ALE.
+    min_bin_points
+        Minimum number of points each discretized interval should contain to ensure more precise
+        ALE estimation.
+
+    Returns
+    -------
+    q
+        Array of quantiles of the input values.
+    ale
+        ALE values for each feature at each of the points in q.
+    ale0
+        The constant offset used to center the ALE curves.
+
+    """
+    q, _ = adaptive_grid(X[:, feature], min_bin_points)
+
+    # find which interval each observation falls into
+    indices = np.searchsorted(q, X[:, feature], side="left")
+    indices[indices == 0] = 1  # put the smallest data point in the first interval
+    interval_n = np.bincount(indices)  # number of points in each interval
+
+    # predictions for the upper and lower ranges of intervals
+    z_low = X.copy()
+    z_high = X.copy()
+    z_low[:, feature] = q[indices - 1]
+    z_high[:, feature] = q[indices]
+    p_low = predictor(z_low)
+    p_high = predictor(z_high)
+
+    # finite differences
+    p_deltas = p_high - p_low
+
+    # make a dataframe for averaging over intervals
+    concat = np.column_stack((p_deltas, indices))
+    df = pd.DataFrame(concat)
+
+    avg_p_deltas = df.groupby(df.shape[1] - 1).mean().values  # groupby indices
+
+    # accummulate over intervals
+    accum_p_deltas = np.cumsum(avg_p_deltas, axis=0)
+
+    # pre-pend 0 for the left-most point
+    zeros = np.zeros((1, accum_p_deltas.shape[1]))
+    accum_p_deltas = np.insert(accum_p_deltas, 0, zeros, axis=0)
+
+    # mean effect, R's `ALEPlot` and `iml` version (approximation per interval)
+    ale0 = (((accum_p_deltas[:-1, :] + accum_p_deltas[1:, :]) / 2) * interval_n[1:, np.newaxis]) \
+               .sum(axis=0) / interval_n.sum()
+
+    # crude approximation (assume datapoints on interval endpoints)
```

Comment:
This is just for future reference, I used the same version the authors used to get exact numbers, afterwards I realized you can be more precise using information of the quantiles (instead of just averaging the midpoints). The crude estimate is not useful but it's the first thing you could think of if approached this.